File: 백준/Gold/1865. 웜홀/웜홀.cc
#include <cstdio>
#include <vector>

using namespace std;

int main()
{
    using FEdge = pair<int, int>;
    constexpr int INF = 1e9;
    
    int TC; 
    int N, M, W, S, E, T;
    vector<vector<FEdge>> Graph;
    vector<int> Distances;
    
    auto UpdateDistances = [&]()
    {
        bool bHasBeenUpdated = false;
        for(int SrcNode = 1; SrcNode <= N; ++SrcNode)
        {
            for(const auto& [DstNode, Time] : Graph[SrcNode])
            {
                if(Distances[SrcNode] + Time < Distances[DstNode])
                {
                    Distances[DstNode] = Distances[SrcNode] + Time;
                    bHasBeenUpdated = true;
                }
            }
        }
        return bHasBeenUpdated;
    };
    
    auto BellmanFord = [&]()
    {
        for(int Repeat = 0; Repeat < N; ++Repeat)
        {
            if(UpdateDistances() && Repeat == N - 1)
            {
                return true;
            }
        }
        return false;
    };
    
    scanf(" %d", &TC);
    while(TC--)
    {
        scanf(" %d %d %d", &N, &M, &W);
        Graph.assign(N + 1, {});
        Distances.assign(N + 1, INF);
        
        for(int i = 0; i < M; ++i)
        {
            
            scanf(" %d %d %d", &S, &E, &T);
            Graph[S].emplace_back(E, T);
            Graph[E].emplace_back(S, T);
        }
        
        for(int i = 0; i < W; ++i)
        {
            scanf(" %d %d %d", &S, &E, &T);
            Graph[S].emplace_back(E, -T);
        }
        
        if(BellmanFord())
        {
            printf("YES\n");
        }
        else
        {
            printf("NO\n");
        }
    }
    
    return 0;
}

File: 백준/Gold/1865. 웜홀/웜홀.py
input = open(0).readline
INF = 10**9

def bellman_ford(G: list, d: list, N: int):
    for _ in range(N - 1):
        for u, v, w in G:
            if d[u] + w < d[v]:
                d[v] = d[u] + w
                
    
    for u, v, w in G:
        if d[u] + w < d[v]:
            return False
            
    return True


for _ in range(TC := int(input())):
    N, M, W = map(int, input().split())
    graph = []
    
    for _ in range(M):
        S, E, T = map(int, input().split())
        graph.append((S, E, T))
        graph.append((E, S, T))
        
    for _ in range(W):
        S, E, T = map(int, input().split())
        graph.append((S, E, -T))
    
    distances = [INF] * (N + 1)
    result = bellman_ford(graph, distances, N)
    print('NO') if result else print('YES')